### ISSUE TITLE
CSRF verification

### ISSUE THREAD

**Issue by keiosweb** [octobercms/library#105]
Currently CSRF verification middleware is commented out in Kernel class and I was unable to find any other place in code doing CSRF checks. Is October using CSRF protection at all?

**Comment by daftspunk** [comment]
Have you tried uncommenting it?

**Comment by keiosweb** [comment]
Uncommenting doesn't make sense, FQCN points to laravel 5's App namespace, which does not exist in October, so kernel crashes with class not found exception, obviously. Please implement this middleware.

**Comment by keiosweb** [comment]
Actually, you could simply change:

     // 'App\Http\Middleware\VerifyCsrfToken',

to:

    'Illuminate\Foundation\Http\Middleware\VerifyCsrfToken',

in `October\Rain\Foundation\Http\Kernel` and it should work.

**Comment by pfrend** [comment]
You can also supply your own middleware by registering your own kernel into the container.

In your Plugin->register() method:

````
//rebind kernel to plugin specific
        \App::singleton(
            'Illuminate\Contracts\Http\Kernel',
            'Acme\YourPlugin\Http\Kernel'
        );
````

Then just register whatever classes you want as middleware in your own Kernel class.
I'm running basically a bare-bones Laravel app as a plugin this way, by just rebinding what I need. That way my user facing app / front end is just normal Laravel but the backend model management is handled by October.

**Comment by keiosweb** [comment]
First: why would I do such thing? Kernel is instantiated long before PluginManager registers any plugins, so this simply won't work. Test this, throw an exception in middleware used by your rebound Kernel class. Do you even application lifecycle?

Second: this is a global issue. Right now October has CSRF protection disabled and don't tell me that this is a sensible default.

**Comment by daftspunk** [comment]
It shouldn't be commented out, we should uncomment it and perform testing to make sure AJAX still works when used outside of a form with the CSRF token.

**Comment by pfrend** [comment]
As to the registration, you're clearly correct, by the time you rebind the kernel the pipeline has already executed. I'm only overriding the $routeMiddleware so wasn't affected by that.

I was merely offering a suggestion to get more control over the Kernel. I agree that CSRF should be turned on by default either way.

However, if wanting control over the global $middleWare you could bind directly to your own Kernel class in bootstrap/app.php instead (with an update to the main composer file).

That's the only way I can think of to be able to provide your own VerifyCsrfToken implementation, like Laravel offers out of the box. The reason I mention it is the developer might want control over when the middleware is applied (e.g. for API's), which is why Laravel offers the CSRF middleware in the app namespace in the first place.

Please let me know if you've got any better ideas here to achieve this level of control.

**Comment by keiosweb** [comment]
I've talked with daftspunk on IRC about this and since bootstrap files are not overwritten on updates best solution for custom builds is to simply use custom kernel class extending October's kernel class.

Alternative would be, as I said before, to read middleware stack configuration from config file such as /config/kernel.php.

@daftspunk: first thing would be to decide what particular behavior for CSRF verification you want in October. My choice would be to enforce CSRF on all POST/DELETE/PUT/PATCH requests and you wouldn't have to modify Otwell's `VerifyCsrfToken` middleware at all to achieve that. What you would have to do would be to export token to every page of CMS, possibly as meta tag. Laravel 5 allows use of _token form field and X-CSRF-TOKEN header, so you could implement this in framework.js. 

Alternatively, you could implement `VerifyCsrfToken` in such way, that it would verify CSRF token only if it was present in POST/DELETE/PUT/PATCH, but this is insecure and generally bad idea.

**Comment by lbialy** [comment]
Bump.

**Comment by keiosweb** [comment]
Bump for security issue. How this hasn't got maximum attention is beyond me.

Just enable 'Illuminate\Foundation\Http\Middleware\VerifyCsrfToken' middleware in October\Rain\Foundation\Http\Kernel's $middleware field. This middleware sets XSRF-TOKEN cookie, you can read that cookie in framework.js and inject it as X-XSRF-TOKEN header.

Problem solved.

**Comment by Mr118** [comment]
@up Please do that, it's quite creepy to build apps with October without CSRF attacks protection.

**Comment by daftspunk** [comment]
[I have added CSRF protection to the back-end](https://github.com/octobercms/october/commit/d9dfa21372049c5ffb5a58ea2715d8733ba053f0) ready for the next build. I am still deciding what to do about the front-end, perhaps a similar solution can be implemented.

The Laravel middleware executes so early there is no possible way to programatically disable the CSRF protection. The solution is to hard code URL addresses in the App middleware... this is, for lack of a nicer word, clunky and undesirable.

These instances should be protected:

* Standard postback
* AJAX submission
* Laravel-based route

These instances **should not** be protected:

* Laravel-based API
* CMS page API
* Back-end Controller API

Disabling the CSRF protection in the backend is as simple as setting `$this->useSecurityToken` to `FALSE` in the backend controller.

Disabling for the frontend can use an approach like this too, or we can explore the idea of building a plugin that implements Laravel's middleware to the CMS controller, introduces an interface that can define "safe" URLs, save the definitions to a meta file on desk, and have the middleware use it as an exclusion list (not exactly my cup of tea). 

Disabling for the Laravel-based API would not be required. We could make Laravel-based routes specify `csrf` route middleware for non-API based routing, i.e. regular pages.

**Comment by keiosweb** [comment]
CMS page-based API sounds like absolutely horrible idea in any case and I really can't think of of any use case that would prefer to use such solution.

Since you don't seem to care about what goes where and have no issues with creating procedural spaghetti codebase (like CSRF protection as a controller method), you could simply copy and paste same solution into Page class or into that dynamic controller class created with page functions. Whatever works.

**Comment by daftspunk** [comment]
Good idea or not, it has to be possible, because somewhere someday someone will want to do it. The code is in the controller for now until we find a better home for it.

**Comment by DanielHitchen** [comment]
Hi guys, I know you're on with this, but just an additional comment - I get the following error whenever I try to install a plugin via the "Install Plugin" search field:
```
october-min.js?v272:5065
Uncaught TypeError: Cannot set property 'X-CSRF-TOKEN' of undefined
```
I've done as @daftspunk suggested & tried changing ```$this->useSecurityToken``` to ```FALSE``` in the backend controller, but it's had no effect - any suggestions on how to suppress this error until the CSRF issue is resolved, as currently I'm unable to search/install plugins via the "Install Plugin" search field? I'm currently on Build 272.

On a side note, I'm also getting a "Security Token Invalid" error whenever I update to a new build - is this related?

**Comment by daftspunk** [comment]
Fixed above issue in 56caa1db3d338df88f35d75f3a5e0348cec2e207.

Please move further discussions to a new issue.
